compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp

```diff
@@ -427,6 +426,18 @@ extern "C" void LLVMRustSetFastMath(LLVMValueRef V) {
   }
 }
 
+// Enable fast-math flags which do not let operations return poison.
+//
+// https://llvm.org/docs/LangRef.html#fast-math-flags
+extern "C" void LLVMRustSetAlgebraicMath(LLVMValueRef V) {
+  if (auto I = dyn_cast<Instruction>(unwrap<Value>(V))) {
+    I->setHasAllowReassoc(true);
+    I->setHasAllowContract(true);
+    I->setHasAllowReciprocal(true);
+    I->setHasNoSignedZeros(true);
```

Comment:
What about `afn` (Approximate functions)? It doesn't poison but isn't mentioned here.

Does the word "algebraic" have a specific meaning here?

Reply:
> What about `afn` (Approximate functions)? It doesn't poison but isn't mentioned here.

Sure, why not. I'll add it.

> Does the word "algebraic" have a specific meaning here?

I didn't want to just defang the existing intrinsics because there are some optimizations which rely on assuming that NaN/Inf do not occur. So I needed to come up with a new name, and the way I think about these intrinsics is that unlike IEEE float operations, they permit the usual algebraic transformations. Things like `a + (b + b) = (a + b) + c` and `a / b = a * (1 / b)`.

I don't think that the name is perfect, and I'd be happy to see someone suggest a better name, but @orlp seems perfectly happy calling them "algebraic".

Reply:
> they permit the usual algebraic transformations. Things like `a + (b + b) = (a + b) + c` and `a / b = a * (1 / b)`

That would be a great explanation to put in a comment somewhere :)

Reply:
> > What about afn (Approximate functions)? It doesn't poison but isn't mentioned here.

> Sure, why not. I'll add it.

Please don't. Replacing functions by their approximations isn't an algebraically justified optimization.

Reply:
:thinking: And in any case, I don't think it would matter for these intrinsics.